MES-4173 - Refactoring cat b fault summary code

TomBillingtonUK opened the pull request:
## Description
- Created a new function which can deal with counting Driving Faults, Serious Faults and Dangerous Faults
- Created a new function for working out Vehicle Checks Faults for Cat B 
- Created a new type FaultSummary and started to use it across the Cat B Journey

Tests haven't been updated as the functionality hasn't changed

## Checklist

- [x] PR title includes the JIRA ticket number
- [x] Branch is rebased against the latest develop
- [ ] Code has been tested manually
- [ ] PR link added to JIRA ticket
- [ ] One review from each scrum team
- [ ] Squashed commit contains the JIRA ticket number

## Screenshots (optional)

TomBillingtonUK committed:
Refactoring

TomBillingtonUK committed:
Started to use FaultSummary in the App

TomBillingtonUK committed:
rename

TomBillingtonUK committed:
Fixed broken sort test

rossfauldsbjss reviewed (approved):
looks good

ammarhaiderbjss commented on src/providers/fault-list/fault-list.ts:
```diff
@@ -48,100 +41,67 @@ export class FaultListProvider {
     return [
-      ...this.getDrivingFaults(data.drivingFaults),
-      ...this.getManoeuvreFaults(data.manoeuvres, CompetencyOutcome.DF),
+      ...this.getComptencyFaults(data.drivingFaults),
```
Typo spelling, "Competency"

ammarhaiderbjss commented on src/shared/models/fault-marking.model.ts:
```diff
@@ -20,3 +20,5 @@ export enum CommentSource {
   UNCOUPLE_RECOUPLE = 'uncoupleRecouple',
 }
+
+export type FaultSummary = Competency & MultiFaultAssignableCompetency & CommentedCompetency;
```
nice!

ammarhaiderbjss commented:
@Cavalier07 Have we refactored any logic here are are we just moving existing stuff around?

ammarhaiderbjss reviewed (approved):
Looks like a nice refactor! Might suggest a further improvement to add a `CompetencyIdentifiers` enum where we can hold all the strings that we do comparison checks on, e.g. "controlFault", "controlledStop" and "eyesightTest". Could also include suffixes as well for things like `endsWith(key, 'Fault')` and `key.endsWith('Comments'))`

TomBillingtonUK committed:
Fixed Typo

TomBillingtonUK commented:
> @Cavalier07 Have we refactored any logic here are we just moving existing stuff around?

I've moved the logic around and simplified some of it. There are 2 new methods (replacing 6 we used to have) but the strucutre and logic is very similar

TomBillingtonUK committed:
Added CompetencyIdentifiers Enum

TomBillingtonUK commented:
> Looks like a nice refactor! Might suggest a further improvement to add a `CompetencyIdentifiers` enum where we can hold all the strings that we do comparison checks on, e.g. "controlFault", "controlledStop" and "eyesightTest". Could also include suffixes as well for things like `endsWith(key, 'Fault')` and `key.endsWith('Comments'))`

Done

ammarhaiderbjss commented on src/shared/models/fault-marking.model.ts:
```diff
@@ -22,3 +22,11 @@ export enum CommentSource {
 export type FaultSummary = Competency & MultiFaultAssignableCompetency & CommentedCompetency;
+
+export enum CompetencyIdentifiers {
+  COMMENTS = 'Comments',
```
Sorry I know its pedantic but can we call this `COMMENTS_SUFFIX` and the fault one `FAULT_SUFFIX` just for clarity.

TomBillingtonUK replied:
done

TomBillingtonUK committed:
adding a suffix
